### PR TITLE
Add support for verifying Intel TDX quotes

### LINKED ISSUE #5824: Add support for verifying Intel TDX quotes
The remote attestation quotes for TDX extend the existing quotes used for SGX/DCAP. There are additional fields and structures for the quotes and TCB infos, but it should be possible to extend the existing quote verification logic.

### PR DESCRIPTION
**Pull request by kostko**
Fixes #5824

### CODE REVIEW

**Review by peternose — COMMENTED** [review]
I think that our code should have a very similar structure as QVL code from Intel. This way, changes can be applied and new versions added more easily. In Rust, I think that we complicated things a lot and it is actually very hard to follow if all verifications were done. In Go, it's a bit easier, even though comments like `/// 4.1.2.4.13` were not copied from the Intel's code, but should be.

File: runtime/src/common/sgx/pcs/quote.rs [hunk 1]
```diff
@@ -0,0 +1,914 @@
...
+    }
+}
+
+/// An enclave quote.
+#[derive(Debug)]
+pub struct Quote<'a> {
+    raw: Cow<'a, [u8]>,
+    header: QuoteHeader<'a>,
+    report_body: Cow<'a, [u8]>,
+    signature: Cow<'a, [u8]>,
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/common/sgx/pcs/quote.rs:160]: Well, in this case `enclave_report` (used previously) would be incorrect as for TDX the reports are TD reports not enclave reports. So `report_body` seems more general.

File: runtime/src/common/sgx/pcs/certificates.rs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+use std::ffi::CString;
+
+use mbedtls::{alloc::List as MbedtlsList, x509::certificate::Certificate};
+
+/// Intel's PCS signing root certificate (PEM-encoded).
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/common/sgx/pcs/certificates.rs:5]: I don't see how this is more clear? The certificate is related to the Provisioning Certification Service (PCS) not only SGX?

File: runtime/src/common/sgx/pcs/quote.rs [hunk 3]
```diff
@@ -0,0 +1,914 @@
...
+            .map(|v| LittleEndian::read_u16(&v))?;
+        match version {
+            QUOTE_VERSION_3 => {
+                // Version 3 (SGX-ECDSA).
+                let att_key_type = quote
+                    .take_prefix(mem::size_of::<u16>())
+                    .map(|v| LittleEndian::read_u16(&v))?;
+                let attestation_key_type = AttestationKeyType::from_u16(att_key_type)
+                    .ok_or(Error::UnsupportedAttestationKeyType)?;
+                let reserved = quote
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/common/sgx/pcs/quote.rs:180]: In the spec it says this is reserved for v3. Where did you see that this is "tee type" for v3? Only in v4 quotes this is "tee type".

File: go/common/node/tee.go [hunk 4]
```diff
@@ -27,6 +27,9 @@ type TEEFeaturesSGX struct {
 
 	// DefaultMaxAttestationAge is the default maximum attestation age (in blocks).
 	DefaultMaxAttestationAge uint64 `json:"max_attestation_age,omitempty"`
+
+	// TDX is a feature flag specifying whether support for TDX is enabled.
+	TDX bool `json:"tdx,omitempty"`
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/node/tee.go:32]: I would rather not because TEE features are already available in the place where this needs to be verified and getting `FeatureVersion` there would require more changes.

File: go/common/sgx/pcs/mock.go [hunk 5]
```diff
@@ -100,7 +100,7 @@ func NewMockQuote(rawReport []byte) ([]byte, error) {
 	var header [quoteHeaderLen]byte
 	binary.LittleEndian.PutUint16(header[0:], 3)                                // Version.
 	binary.LittleEndian.PutUint16(header[2:], uint16(AttestationKeyECDSA_P256)) // Attestation key type.
-	binary.LittleEndian.PutUint32(header[4:], teeTypeSGX)                       // TEE type.
+	binary.LittleEndian.PutUint32(header[4:], 0)                                // Reserved.
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/pcs/mock.go:103]: I was reading the v3 spec where it says this is reserved. Where did you see that this is "tee type" in v3?

File: go/common/sgx/pcs/policy.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	// packages and platform instances are blocked.
+	FMSPCBlacklist []string `json:"fmspc_blacklist,omitempty" yaml:"fmspc_blacklist,omitempty"`
+
+	// TDX is an optional TDX-specific policy. In case this is nil, TDX quotes are disallowed.
+	TDX *TdxQuotePolicy `json:"tdx,omitempty" yaml:"tdx,omitempty"`
+}
+
+// TdxQuotePolicy is the TDX-specific quote policy.
+type TdxQuotePolicy struct {
+	// AllowedTdxModules are the allowed TDX modules. Empty to allow ANY Intel-signed module.
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/pcs/policy.go:29]: Currently only Intel-signed modules exist. Why is this scary? Note that this is an additional restriction as all TDX modules are already verified against the TCB infos.

File: runtime/src/common/sgx/pcs/certificates.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (peternose) [reply runtime/src/common/sgx/pcs/certificates.rs:5]: Maybe it is not.  Ignore.

File: runtime/src/common/sgx/pcs/quote.rs [hunk 7]
```diff
@@ -0,0 +1,914 @@
...
+                    .take_prefix(mem::size_of::<u16>())
+                    .map(|v| LittleEndian::read_u16(&v))?;
+                let pce_svn = quote
+                    .take_prefix(mem::size_of::<u16>())
+                    .map(|v| LittleEndian::read_u16(&v))?;
+                let qe_vendor_id = quote.take_prefix(QE_VENDOR_ID_LEN)?;
+                let user_data = quote.take_prefix(QE_USER_DATA_LEN)?;
+
+                // Ensure correct QE vendor and attestation key type.
+                if *qe_vendor_id != QE_VENDOR_ID_INTEL[..] {
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/common/sgx/pcs/quote.rs:197]: Will move.

File: go/common/sgx/pcs/mock.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (peternose) [reply go/common/sgx/pcs/mock.go:103]: You are right, was looking at Intel's implementation.

File: runtime/src/common/sgx/pcs/quote.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (peternose) [reply runtime/src/common/sgx/pcs/quote.rs:160]: You are right, `report_body` is fine for `enclave_report` and  `td_report`. Since this was inspired by Fortanix, ignore this comment (and some other comments as well).

File: go/common/sgx/pcs/policy.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (peternose) [reply go/common/sgx/pcs/policy.go:29]: I thought this was scary because if a user wanted to restrict modules to only one and forgot to add it to the list, he would mistakenly allow all Intel-signed modules. Which is not what you would think first when you see an empty access/white list.

File: runtime/src/common/sgx/pcs/quote.rs [hunk 8]
```diff
@@ -0,0 +1,914 @@
...
+        let sig = get_ecdsa_sig_der(self.qe_signature())?;
+        let mut hash = [0u8; 32];
+        Md::hash(hash::Type::Sha256, self.qe_report(), &mut hash)
+            .map_err(|err| Error::Other(err.into()))?;
+        let mut pck_pk = Pk::from_public_key(pck_pk).map_err(|err| Error::Other(err.into()))?;
+        pck_pk
+            .verify(mbedtls::hash::Type::Sha256, &hash, &sig)
+            .map_err(|_| Error::VerificationFailed("QE report signature is invalid".to_string()))?;
+
+        let mut qe_report = Vec::with_capacity(SgxReport::UNPADDED_SIZE);
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/common/sgx/pcs/quote.rs:735]: Where would you move this part? It is needed below.

File: runtime/src/common/sgx/pcs/quote.rs [hunk 9]
```diff
@@ -0,0 +1,914 @@
...
+    /// Verify signature against quote using the attestation public key.
+    ///
+    /// The passed `data` must cover the Quote Header and the Report Data.
+    pub fn verify_quote_signature(&'a self, data: &[u8]) -> Result<&'a Self, Error> {
+        let sig = get_ecdsa_sig_der(self.signature())?;
+        let mut hash = [0u8; 32];
+        Md::hash(hash::Type::Sha256, data, &mut hash).map_err(|err| Error::Other(err.into()))?;
+
+        let mut pk = self.attestation_pk()?;
+        pk.verify(mbedtls::hash::Type::Sha256, &hash, &sig)
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/common/sgx/pcs/quote.rs:488]: > For clarity, I would also move `sig` before `pk`.

Isn't this already the case?

File: go/common/sgx/pcs/policy.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/pcs/policy.go:29]: The idea is that an empty policy would allow all Intel-signed modules. I guess we could explicitly have an `Option<Vec<_>>`? Or do you have any other suggestions.

File: runtime/src/common/sgx/pcs/tcb.rs [hunk 10]
```diff
@@ -0,0 +1,686 @@
...
+            let tdx_module_version = tdx_comp_svn[1];
+            if tdx_module_version >= 1 {
+                // In order to determine TCB status of TDX module, find a matching TDX Module
+                // Identity (in tdxModuleIdentities array of TCB Info) with its id set to
+                // "TDX_<version>" where <version> matches the value of TEE TCB SVN at index 1. If a
+                // matching TDX Module Identity cannot be found, fail.
+                let tdx_module = self
+                    .tdx_module_identities
+                    .iter()
+                    .find(|tm| tm.id == format!("TDX_{:02}", tdx_module_version))
```
Comment (peternose) [inline runtime/src/common/sgx/pcs/tcb.rs:274]: Same here.

File: runtime/src/common/sgx/pcs/quote.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (peternose) [reply runtime/src/common/sgx/pcs/quote.rs:761]: I would move with together with code below.

File: go/common/sgx/pcs/policy.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (peternose) [reply go/common/sgx/pcs/policy.go:29]: Keep it as it is. Since we allow only Intel signed modules and not all, maybe it is not that scary.

**Review by peternose — APPROVED** [review]
Looks good, have no major comments, just few minor (address them as you like).

File: go/common/sgx/pcs/tcb.go [hunk 11]
```diff
@@ -319,6 +345,50 @@ func (ti *TCBInfo) getTCBLevel(
...
+			// (in tdxModuleIdentities array of TCB Info) with its id set to "TDX_<version>" where
+			// <version> matches the value of TEE TCB SVN at index 1. If a matching TDX Module
+			// Identity cannot be found, fail.
+			idx := slices.IndexFunc(ti.TDXModuleIdentities, func(tm TDXModuleIdentity) bool {
+				return tm.ID == fmt.Sprintf("TDX_%02d", tdxModuleVersion)
+			})
+			if idx < 0 {
+				return nil, fmt.Errorf("pcs/tcb: TDX module not supported")
+			}
+			// Otherwise, for the selected TDX Module Identity go over the sorted collection of TCB
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/pcs/tcb.go:365]: This is quoted from the Intel spec.

File: go/common/sgx/pcs/tcb.go [hunk 12]
```diff
@@ -376,25 +446,46 @@ type TCBLevel struct {
...
-	//    assigned to this TCB level. Otherwise, move to the next item on TCB Levels list.
-	if tl.TCB.PCESVN < pcesvn { //nolint:gosimple // Explicit is better.
+	//    assigned to this TCB level (in case of SGX) or go to c (in case of TDX). Otherwise, move
+	//    to the next item on TCB Levels list.
+	if tl.TCB.PCESVN < pcesvn {
 		return false
 	}
 
+	if tdxCompSvn != nil {
+		// c) Compare SVNs in TEE TCB SVN array retrieved from TD Report in Quote (from index 0 to
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/pcs/tcb.go:470]: This is quoted from the Intel spec to show where a particular verification step is implemented.

File: go/common/sgx/pcs/report.go [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+	}
+}
+
+// Raw returns the raw report body.
+func (r *TdxReport) Raw() []byte {
+	return r.raw
+}
+
+// TdxAttributes are the TDX TD attributes.
+type TdxAttributes uint64
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/pcs/report.go:203]: Because these are bitflags and having them as `uint64` makes them easier to test.

File: go/common/sgx/pcs/report.go [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+	if attrs.Contains(TdxAttributeReserved) {
+		return fmt.Errorf("pcs/quote: malformed TDX attributes in report body")
+	}
+
+	*a = attrs
+	return nil
+}
+
+// Contains returns value of given flag attribute of the Report.
+func (a TdxAttributes) Contains(flag TdxAttributes) bool {
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/pcs/report.go:235]: False, even though the use case is to specify a single flag. If you specify multiple then all of the flags need to be set.

File: go/common/sgx/pcs/quote.go [hunk 15]
```diff
@@ -142,65 +138,221 @@ func (q *Quote) Verify(policy *QuotePolicy, ts time.Time, tcb *TCBBundle) (*sgx.
...
+// UnmarshalBinary decodes QuoteHeaderV3 from a byte array.
+func (qh *QuoteHeaderV3) UnmarshalBinary(data []byte) error {
 	if len(data) != quoteHeaderLen {
 		return fmt.Errorf("pcs/quote: invalid quote header length")
 	}
 
-	qh.Version = binary.LittleEndian.Uint16(data[0:])
-	if qh.Version != 3 {
-		return fmt.Errorf("pcs/quote: unsupported quote version %d", qh.Version)
+	qh.attestationKeyType = AttestationKeyType(binary.LittleEndian.Uint16(data[2:]))
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/pcs/quote.go:236]: It is verified in quote unmarshal against supported key types.

File: .buildkite/rust/coverage.sh [hunk 16]
```diff
@@ -41,6 +41,7 @@ export RUSTFLAGS="-C target-feature=+aes,+ssse3"
 
 # Make sure we can run unit tests for production enclaves.
 unset OASIS_UNSAFE_ALLOW_DEBUG_ENCLAVES
+unset OASIS_UNSAFE_SKIP_AVR_VERIFY
```
Comment (peternose) [inline .buildkite/rust/coverage.sh:44]: a

File: go/common/sgx/pcs/tcb.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (peternose) [reply go/common/sgx/pcs/tcb.go:366]: It's a lot of lines to read, giving you no extra info. You can leave it.

File: go/common/sgx/pcs/report.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (peternose) [reply go/common/sgx/pcs/report.go]: So this code needs to be fixed as it currently returns `true`?

File: go/common/sgx/pcs/report.go [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+// UnmarshalBinary decodes TdxReport from a byte array.
+func (r *TdxReport) UnmarshalBinary(data []byte) error {
+	if len(data) < reportBodyTdxLen {
+		return fmt.Errorf("pcs/quote: invalid report length")
+	}
+
+	var teeTcbSvn [16]byte
+	copy(teeTcbSvn[:], data[0:])
+	for i, v := range teeTcbSvn {
+		r.teeTcbSvn[i] = int32(v)
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/pcs/report.go:138]: Agree, will change.

File: go/common/sgx/pcs/report.go [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+
+// ReportData is the user-supplied report data.
+func (r *TdxReport) ReportData() []byte {
+	return r.reportData[:]
+}
+
+// AsEnclaveIdentity converts the report body into an enclave identity.
+func (r *TdxReport) AsEnclaveIdentity() sgx.EnclaveIdentity {
+	var zeroMrSigner sgx.MrSigner
+	// TODO: Change the EnclaveIdentity structure to allow specifying all the different things.
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/pcs/report.go:171]: Not in this PR :)

File: go/common/sgx/pcs/report.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/pcs/report.go]: Oh, you're right, let me fix that.

[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/pcs/report.go]: Fixed.
